tools/perfscale-audit/metric-client/metric-client.go

```diff
@@ -0,0 +1,169 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package metricclient
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"log"
+	"math"
+	"net/http"
+	"time"
+
+	audit_api "kubevirt.io/kubevirt/tools/perfscale-audit/api"
+
+	api "github.com/prometheus/client_golang/api"
+	apiv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+)
+
+type transport struct {
+	transport http.RoundTripper
+	userName  string
+	password  string
+	token     string
+}
+
+func (a transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if a.userName != "" {
+		req.SetBasicAuth(a.userName, a.password)
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", a.token))
+	return a.transport.RoundTrip(req)
+}
+
+type MetricClient struct {
+	client apiv1.API
+	cfg    *audit_api.InputConfig
+}
+
+func NewMetricClient(cfg *audit_api.InputConfig) (*MetricClient, error) {
+
+	url := cfg.PrometheusURL
+	token := cfg.PrometheusBearerToken
+	userName := cfg.PrometheusUserName
+	password := cfg.PrometheusPassword
+	tlsVerify := cfg.PrometheusVerifyTLS
+
+	apiCfg := api.Config{
+		Address: url,
+		RoundTripper: transport{
+			transport: &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: tlsVerify}},
+			token:     token,
+			userName:  userName,
+			password:  password,
+		},
+	}
+	c, err := api.NewClient(apiCfg)
+	if err != nil {
+		return nil, err
+	}
+	api := apiv1.NewAPI(c)
+
+	_, err = api.Config(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+	log.Print("Established connection with prometheus endpoint.")
+
+	return &MetricClient{client: api, cfg: cfg}, nil
+}
+
+func (m *MetricClient) query(query string) (model.Value, error) {
+	log.Printf("Makeing query [%s]", query)
+	val, _, err := m.client.Query(context.TODO(), query, *m.cfg.EndTime)
+	if err != nil {
+		return val, err
+	}
+	return val, nil
+
+}
+
+type metric struct {
+	labels    map[string]string
+	value     float64
+	timestamp time.Time
+}
+
+func parseVector(value model.Value) ([]metric, error) {
+	var metrics []metric
+
+	data, ok := value.(model.Vector)
+	if !ok {
+		return metrics, fmt.Errorf("unexpected format %s, expected vector", value.Type().String())
+	}
+	for _, v := range data {
+		m := metric{
+			labels: make(map[string]string),
+		}
+		for k, v := range v.Metric {
+			m.labels[string(k)] = string(v)
+		}
+		if math.IsNaN(float64(v.Value)) {
+			m.value = 0
+		} else {
+			m.value = float64(v.Value)
+		}
+		m.timestamp = v.Timestamp.Time()
+		metrics = append(metrics, m)
+	}
+	return metrics, nil
+}
+
+func (m *MetricClient) getCreationToRunningTimePercentile(percentile int) (float64, error) {
+	query := fmt.Sprintf(`histogram_quantile(0.%d, rate(kubevirt_vmi_phase_transition_time_from_creation_seconds_bucket{phase="Running"}[%ds]))`, percentile, int(m.cfg.Duration.Seconds()))
```

Comment:
It would be nice to have the metrics defined as constants. For example:
```
const (
        // VMI creation time from object creation to the given phase
        vmiCreationTime = histogram_quantile(0.%d, rate(kubevirt_vmi_phase_transition_time_from_creation_seconds_bucket{phase="%s"}[%ds]))
)
```
When we include other metrics, we will have a list of supported metrics

Reply:
yup, makes sense. I updated this